[docs] add zoom transition docs (#41516)
expo/expo: docs/pages/router/advanced/zoom-transition.mdx

---

> title: Zoom transition

Author: > If the second goal is true from above and we're okay with it being included by LLMs and search engines at this point, then please ignore my comment.

I think this would be ideal. If LLMs can understand that this is SDK 55 feature, only available in canary version then it's ok for the docs to be indexed

> description: Learn how to use the zoom transition to create fluid animations between screens.

Reviewer: ```suggestion
description: Learn how to use the zoom transition to create fluid animations between screens when using Expo Router.
```

Let's add the context for SEO and LLMs.

hasVideoLink: false
---

import { ContentSpotlight } from '~/ui/components/ContentSpotlight';

> > **important** Zoom transition is currently experimental and available in canary releases of `expo-router`.

Reviewer: ```suggestion
> **important** Zoom transition is currently in [alpha](/more/release-statuses/#alpha) and available in canary releases of `expo-router`.
```

We have recently defined Release statuses: https://docs.expo.dev/more/release-statuses/. For consistency, I'd suggest using the term we are expressing in the term.

> > **info** In order to enable zoom transitions, run `enableZoomTransition();` in your `app/_layout.tsx` file.

Reviewer: ```suggestion
> **info** **Tip:** To enable zoom transitions, run `enableZoomTransition()` in your **app/_layout.tsx`** file.
```

<ContentSpotlight
  alt="A video presenting simple gallery with zoom transitions."
  file="expo-router/zoom-transition.mp4"
/>

Zoom transitions provide a fluid animation effect when navigating between screens by zooming from a source element to the destination screen. This feature leverages iOS 18+'s native zoom transition API to create smooth, interactive transitions that make your app feel more polished and responsive.

## Get started

> To implement zoom transitions, you'll use the `Link.AppleZoom` component to mark the source element and optionally `Link.AppleZoom.Target` to specify the target alignment on the destination screen.

Reviewer: ```suggestion
To implement zoom transitions, you need to use the `Link.AppleZoom` component to mark the source element and optionally `Link.AppleZoom.Target` to specify the target alignment on the destination screen.
```

### Basic example

Here's a simple example showing how to create a zoom transition from an image to a preview screen:

```tsx app/_layout.tsx
import { enableZoomTransition } from 'expo-router/internal/utils';

enableZoomTransition();
export default function Layout() {
  ... // your layout code
}
```

> ```tsx app/index.tsx

Reviewer: ````suggestion
Add `Link.AppleZoom` to wrap the source element (`Image`) in your screen component that you want to zoom from:

```tsx app/index.tsx
````

import { View, Text, StyleSheet, Pressable } from 'react-native';
import { Link } from 'expo-router';

export default function HomeScreen() {
  return (
    <View style={styles.container}>
      <Link href="/image" asChild>
        <Link.AppleZoom>
          <Pressable>
            <Image
              source={{ uri: 'https://example.com/image-1.jpg' }}
              style={{ width: 100, height: 200 }}
            />
          </Pressable>
        </Link.AppleZoom>
      </Link>
    </View>
  );
}
```

> ```tsx app/image.tsx

Reviewer: ````suggestion
In the destination screen, define the `Image` component:

```tsx app/image.tsx
````

      <View>{/* Your content */}</View>
    </Link.AppleZoom>
    <Text>Subtitle</Text>
  </Pressable>
</Link>
```

> **info** `Link.AppleZoom` only accepts a single child component. If you need to wrap multiple children, use a `View` or another container component.

### Customizing alignment

You can specify the alignment of the zoomed element on the destination screen by using `Link.AppleZoom.Target` element.

```tsx app/image.tsx
export default function ImageScreen() {
  return (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <Link.AppleZoom.Target>
        <Image source={{ uri: 'https://example.com/image-1.jpg' }} style={{ width: '100%' }} />
      </Link.AppleZoom.Target>
    </View>
  );
}
```

If you need more control over the alignment rectangle, you can pass an `alignmentRect` prop to `Link.AppleZoom`. However, this is normally not necessary if you use `Link.AppleZoom.Target`.

```tsx
<Link.AppleZoom alignmentRect={{ x: 0, y: 0, width: 200, height: 300 }}>
  <Image source={{ uri: 'https://example.com/image-1.jpg' }} style={{ width: 100, height: 150 }} />
</Link.AppleZoom>
```

## Advanced examples

### Gallery with zoom transitions

> Here's a more complex example showing a gallery grid with zoom transitions to detail views:

Reviewer: ```suggestion
Here's a more complex example showing a gallery grid with zoom transitions to detail views. The source screen component (**app/index.tsx**) uses `Link.AppleZoom` to wrap an `Image` component:
```

          />
        </Link.AppleZoom>
        <Text style={{ textAlign: 'center' }}>Photo {index + 1}</Text>
      </Pressable>
    </Link>
  );
}

const styles = StyleSheet.create({
  scrollView: {
    flex: 1,
    backgroundColor: '#fff',
    padding: 16,
  },
  scrollViewContent: {
    justifyContent: 'center',
    flexDirection: 'row',
    flexWrap: 'wrap',
    gap: 16,
  },
  thumbnail: {
    width: 170,
    aspectRatio: 1,
  },
  thumbnailImage: {
    width: '100%',
    height: '100%',
    borderRadius: 8,
  },
});
```

```tsx app/image/[id].tsx
import { Image } from 'expo-image';
import { Link, useLocalSearchParams } from 'expo-router';
import { useMemo } from 'react';
import { StyleSheet, useWindowDimensions, View } from 'react-native';

export default function ImagePage() {
  const params = useLocalSearchParams();
  const index = params.id ? parseInt(params.id as string, 10) : 0;
  const imageSource = IMAGES[index % IMAGES.length];
  const imageSize = {
    width: parseInt(params.width as string, 10),
    height: parseInt(params.height as string, 10),
  };
  const windowDimensions = useWindowDimensions();

>   // Compute the size to fit within the window while maintaining aspect ratio

Reviewer: ```suggestion
  // Compute the size to fit within the window while maintaining aspect ratio.
```

    const minRatio = Math.min(widthRatio, heightRatio);
    return {
      width: imageSize.width * minRatio,
      height: imageSize.height * minRatio,
    };
  }, [imageSize, windowDimensions]);

  return (
    <View style={styles.container}>
      <Link.AppleZoomTarget>
        <View style={{ ...computedSize }}>
          <Image source={imageSource} style={styles.image} />
        </View>
      </Link.AppleZoomTarget>
    </View>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: '#fff', justifyContent: 'center', alignItems: 'center' },
  image: { width: '100%', height: '100%' },
});
```

## Platform support

Zoom transitions is only available on iOS 18 and later. On older iOS versions or other platforms, the component will render normally without the zoom animation effect.

The zoom transition components automatically detect platform support and gracefully degrade to standard navigation on unsupported platforms.

## Known limitations

### Using zoom transition with headers

We recommend avoiding the use of zoom transitions when navigating between screens that have a header (navigation bar).

> There are known issues with the native iOS zoom transition API that can lead to visual glitches or unexpected behavior when headers are involved.

Reviewer: ```suggestion
We recommend avoiding the use of zoom transitions when navigating between screens that have a header (navigation bar). There are known issues with the native iOS zoom transition API that can lead to visual glitches or unexpected behavior when headers are involved.
```
